src/lib/parsers/ChatInputParser.ts

```diff
@@ -0,0 +1,59 @@
+import { type CommandInteraction, type CommandInteractionOption } from 'discord.js';
+import { Option, Result } from '@sapphire/result';
+import { type Parameter } from '@sapphire/lexure';
+
+export class ChatInputParser {
+	public position: number = 0;
+
+	public constructor(public interaction: CommandInteraction) {}
+
+	public get finished(): boolean {
+		return this.position === this.interaction.options.data.length;
+	}
+
+	public reset(): void {
+		this.position = 0;
+	}
+
+	public save(): number {
+		return this.position;
+	}
+
+	public restore(state: number): void {
+		this.position = state;
+	}
+
+	public async singleParseAsync<T, E>(
+		predicate: (arg: CommandInteractionOption) => Promise<Result<T, E>>,
+		useAnyways?: boolean
+	): Promise<Result<T, E | null>> {
+		if (this.finished) return Result.err(null);
+
+		const result = await predicate(this.interaction.options.data[this.position]);
```

Comment:
This should not rely on any magic position. That's why we pass in a name now in the arguments (`args.pick({ name: 'owo', type: 'string' })` -> find the argument with name `owo` (return if missing), parse as string (we can skip a LOT of parsing depending how we parse arguments. For example, with users, we know the value is the id, but we may get it from the resolved field!)

I'd say overall, if you want to implement a parser-esque to the Lexer interface, you'd just want to track what args were found and "consumed", and reject if it was already used.

something like Record<argName, boolean>, and save/restore would store the object as is, and make a clone of it